midend/predication.cpp

```diff
@@ -15,13 +15,22 @@ limitations under the License.
 */
 
 #include "predication.h"
+#include "frontends/p4/cloner.h"
 
 namespace P4 {
 
+const IR::Expression* Predication::clone(const IR::Expression* expression) {
+    ClonePathExpressions cloner;
+    auto result = expression->apply(cloner)->to<IR::Expression>();
+    CHECK_NULL(result);
+    return result;
+}
+
 const IR::Node* Predication::postorder(IR::AssignmentStatement* statement) {
     if (!inside_action || ifNestingLevel == 0)
         return statement;
-    auto right = new IR::Mux(predicate(), statement->right, statement->left);
+
+    auto right = new IR::Mux(predicate(), statement->right, clone(statement->left));
```

Comment:
A comment why do you need to clone this would be helpful.

Reply:
I am afraid I won't have time to add this comment in the next 3 weeks.
The reason is that in the end we generate different code for the two copies of the expression, so we want them to be distinct. This happens in many other passes, not only in this one.